// SPDX-License-Identifier: AGPL-3.0-or-later
/*
 * ZDB a distributed, fault-tolerant database.
 * Copyright (C) 2025 Ahmed Refaat Gadalla Mohamed
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

syntax = "proto3";

package zdb.proto;

enum ErrorCode {
    OK = 0;
    InvalidArg = 1;
    ServiceTemporarilyUnavailable = 2;
    AllServicesUnavailable = 3;
    VersionMismatch = 4;
    Maybe = 5;
    KeyNotFound = 6;
    Timeout = 7;
    NotLeader = 8;
    Internal = 9;
    Cancelled = 10;
    Unknown = 128;
}

message ErrorDetails {
    ErrorCode code = 1;
    string what = 2;
    string key = 3;
    string value = 4;
    uint64 version = 5;
}
